### PR TITLE
[SYNC-1751] Fix packaging of nimbus to include native libs for unittesting.

### PR DESCRIPTION
**Pull request by travis79**
This is a WIP towards getting unit tests to run (in consuming code like the [Nimbus Android-Component](https://github.com/mozilla-mobile/android-components/pull/8261)) and include the local system libs rather than just the android libs.

### CODE REVIEW

File: android/.buildconfig.yaml [hunk 1]
```diff
@@ -2,3 +2,7 @@ libraryVersion: 0.1.0
 groupId: uniffi
 artifactId: nimbus
 description: A uniffi generated android bindings for the Nimbus SDK
+libUrl: https://github.com/mozilla/glean
```
Comment (vladikoff) [inline android/.buildconfig.yaml:5]: Is it correct to reference glean here?

File: android/build.gradle [hunk 2]
```diff
@@ -49,7 +49,27 @@ buildscript {
...
+    // doesn't provide the JNI dispatch libraries in the correct Java resource directories.  I think
+    // what's happening is that @aar type in `implementation` resolves to the @jar type in
+    // `testImplementation`, and that it wins the dependency resolution battle.
+    //
+    // A workaround is to add a new configuration which depends on the @jar type and to reference
+    // the underlying JAR file directly in `testImplementation`.  This JAR file doesn't resolve to
+    // the @aar type in `implementation`.  This works when invoked via `gradle`, but also sets the
+    // correct runtime classpath when invoked with Android Studio's built-in JUnit test runner.
+    // Success!
+    jnaForTest
```
Comment (vladikoff) [inline android/build.gradle:66]: seen that before 👍

File: android/build.gradle [hunk 3]
```diff
@@ -90,18 +137,30 @@ allprojects {
...
         'arm',
         'arm64',
         'x86_64',
         'x86',
-    ]
-
+]
+
+// Generate libs for our current platform so we can run unit tests.
+switch (DefaultPlatform.RESOURCE_PREFIX) {
```
Comment (vladikoff) [inline android/build.gradle:152]: 👍

File: android/.buildconfig.yaml (hunk 1 above)
Comment (travis79) [reply android/.buildconfig.yaml:5]: No, this is copy-pasta, will clean it up!

**Comment by rfk** [comment]
Ref https://github.com/travis79/nimbus-sdk/pull/1 for some more progress on getting these test passing, hope it's helpful

**Comment by travis79** [comment]
I believe that this is working now, thanks to @rfk's help.  Ready for review.

**Review by rfk — COMMENTED** [review]
I tried doing a fresh checkout and `./gradlew publishToMavenLocal` here to see if it built as I'd expect.

The `-forUnitTest` variant seems to build correctly and include the `libnimbus.so` for my local machine:

```
rfk@pomelo:nimbus-sdk(fix-unit-tests)$ unzip -l ~/.m2/repository/uniffi/nimbus-forUnitTests/0.1.0/nimbus-forUnitTests-0.1.0.jar | grep 'libnimbus.so'
3930768  2020-09-21 13:56   linux-x86-64/libnimbus.so                                                                                               
```

But the main `.aar` does not seem to include the compiled rust code:

```
rfk@pomelo:android(fix-unit-tests)$ unzip -l ~/.m2/repository/uniffi/nimbus/0.1.0/nimbus-0.1.0.aar
Archive:  /home/rfk/.m2/repository/uniffi/nimbus/0.1.0/nimbus-0.1.0.aar
Length      Date    Time    Name
---------  ---------- -----   ---
        0  1980-02-01 00:00   proguard.txt
       35  1980-02-01 00:00   R.txt
      299  1980-02-01 00:00   AndroidManifest.xml
   122497  1980-02-01 00:00   classes.jar
---------                     -------
   122831                     4 files                                                             
```

Uncommenting this commented-out line seems to make it build correctly for me:

```
tasks["merge${productFlavor}${buildType}JniLibFolders"].dependsOn(tasks["cargoBuild"])
```

But I assume that it's commented out on purpose. Could you please say more about what was (or wasn't) happening with that line enabled on your machine?

**Comment by rfk** [comment]
For completeness, after uncommenting that line I get:

```
rfk@pomelo:android(fix-unit-tests *)$ unzip -l ~/.m2/repository/uniffi/nimbus/0.1.0/nimbus-0.1.0.aar | grep libnimbus.so
  1083544  1980-02-01 00:00   jni/arm64-v8a/libnimbus.so
   751216  1980-02-01 00:00   jni/armeabi-v7a/libnimbus.so
  1099436  1980-02-01 00:00   jni/x86/libnimbus.so
  1210736  1980-02-01 00:00   jni/x86_64/libnimbus.so                                                                   
```

Which looks about how I'd expect it for a rust component.

File: android/build.gradle [hunk 4]
```diff
@@ -136,7 +202,12 @@ afterEvaluate {
             productFlavor += "${it.name.capitalize()}"
         }
         def buildType = "${variant.buildType.name.capitalize()}"
-        tasks["merge${productFlavor}${buildType}JniLibFolders"].dependsOn(tasks["cargoBuild"])
+        // tasks["merge${productFlavor}${buildType}JniLibFolders"].dependsOn(tasks["cargoBuild"])
+
+        tasks["generate${productFlavor}${buildType}Assets"].dependsOn(tasks["cargoBuild"])
```
[the remark this answers is not in the document]
Comment (travis79) [reply android/build.gradle]: Ahh, that was just an artifact of trying to figure out getting the native libs into the bundle.  I had originally commented that out to see if it was affecting the problem I was having.  I don't think that the next line with the `generate` command shouldn't be necessary either.

**Review by rfk — APPROVED** [review]
Looks good!

As mentioned in Matrix, I'm sad that you had to copy-paste-and-tweak so much guff to make this work, and we should think about pulling this into a gradle plugin at some point.

**Pull request merged**